metaserver/src/main/scala/scala/meta/languageserver/search/SymbolIndex.scala

```diff
@@ -1,256 +1,55 @@
-  }
-
-  /**
-   *
-   * Register this Document to symbol indexer.
-   *
-   * Indexes definitions, denotations and references in this document.
-   *
-   * @param document Must respect the following conventions:
-   *                 - filename must be a URI
-   *                 - names must be sorted
-   */
-  def indexDocument(document: s.Document): Effects.IndexSemanticdb = {
-    val input = Input.VirtualFile(document.filename, document.contents)
-    // what do we put as the uri?
-    val uri = URI.create(document.filename)
-    documentIndex.putDocument(uri, document)
-    document.names.foreach {
-      // TODO(olafur) handle local symbols on the fly from a `Document` in go-to-definition
-      // local symbols don't need to be indexed globally, by skipping them we should
-      // def isLocalSymbol(sym: String): Boolean =
-      // !sym.endsWith(".") &&
-      //     !sym.endsWith("#") &&
-      //     !sym.endsWith(")")
-      // be able to minimize the size of the global index significantly.
-      //      case s.ResolvedName(_, sym, _) if isLocalSymbol(sym) => // Do nothing, local symbol.
-      case s.ResolvedName(Some(s.Position(start, end)), sym, true) =>
-        symbolIndexer.addDefinition(
-          sym,
-          i.Position(document.filename, Some(input.toIndexRange(start, end)))
-        )
-      case s.ResolvedName(Some(s.Position(start, end)), sym, false) =>
-        symbolIndexer.addReference(
-          document.filename,
-          input.toIndexRange(start, end),
-          sym
-        )
-      case _ =>
-    }
-    document.symbols.foreach {
-      case s.ResolvedSymbol(sym, Some(denot)) =>
-        symbolIndexer.addDenotation(
-          sym,
-          denot.flags,
-          denot.name,
-          denot.signature
-        )
-      case _ =>
-    }
-    Effects.IndexSemanticdb
-  }
-
-  /**
-   * Returns false this this document is stale.
-   *
-   * A document is considered stale if it's off-sync with the contents in [[buffers]].
-   */
-  private def isFreshSemanticdb(
-      path: AbsolutePath,
-      document: Document
-  ): Option[Unit] = {
-    val ok = Option(())
-    val s = buffers.read(path)
-    if (s == document.contents) ok
-    else {
-      // NOTE(olafur) it may be a bit annoying to bail on a single character
-      // edit in the file. In the future, we can try more to make sense of
-      // partially fresh files using something like edit distance.
-      notifications.showMessage(
-        l.MessageType.Warning,
-        "Please recompile for up-to-date information"
-      )
-      None
-    }
-  }
-
+trait SymbolIndex {
+  def findSymbol(path: AbsolutePath, line: Int, column: Int): Option[Symbol]
+  def definitionData(symbol: Symbol): Option[SymbolData]
+  def referencesData(symbol: Symbol): List[SymbolData]
+  def indexDependencyClasspath(sourceJars: List[AbsolutePath]): Effects.IndexSourcesClasspath
+  def indexDatabase(document: s.Database): Effects.IndexSemanticdb
 }
```

Comment:
@olafurpg What do you think about turning SymbolIndex into a small trait with a curated list of methods? If you're on board with that, I can submit a pull request with just this change right away.

Reply:
I like it, that's a great first step 👍